3.0 stricter inflector

AD7six opened the pull request:
Following discussions related to #5851 - These changes make the inflector methods more strict/consistent.

E.g.:

```
 $this->assertSame('test_this_thing', Inflector::dasherize('test_this_thing'), 'Should be unchanged');
```

Because dasherize previously wrapped underscore, it could be (ab)used to accept underscored input even if that wasn't desired.

AD7six committed:
Consolidate underscore, dasherize and humanize

All 3 do basically the same thing - so create a method for normalizing a
string, and make all 3 call that same method

AD7six committed:
Allow directly specifying the separator for camelize

AD7six committed:
Use consistent method logic

AD7six committed:
Correct humanize

There isn't any real logic here - it's just a str replace call

jadb commented:
:+1:

ADmad commented:
> Because dasherize previously wrapped underscore, it could be (ab)used to accept underscored input even if that wasn't desired.

The replacement of underscore with dash is intentional. Dasherize  of [rails](http://apidock.com/rails/String/dasherize) and [ember](http://emberjs.com/api/classes/Ember.String.html#method_dasherize) also replace `_` with `-`.

We could add an extra param to it to control replacement of underscore (for use in routes parsing) but by default it should replace as it does now.

ADmad commented:
CS errors.

markstory commented:
I think dasherize should be replacing underscores as well. I think the method wouldn't be doing what its name says if underscores were left behind.

AD7six commented:
I see that that's how [rails acts](http://api.rubyonrails.org/classes/ActiveSupport/Inflector.html#method-i-dasherize) - so fine with me, though to me it's somewhat inconsistent atm:

 * You can feed "this_that" to dasherize and get "this-that".
 * But you _can't_ feed "this-that" to underscore and get "this_that".

Should underscore and dasherize be equally lenient?

I don't think normalize is the right name for that method - any better name?

Also whilst I'm asking [humanize is such a trivial method](https://github.com/cakephp/cakephp/pull/5859/files#diff-19eff47ebbcb312510b116f11192a975R632) is it worth having cache on a str_replace and ucwords call (I doubt it)?

ADmad commented:
I wouldn't mind if underscore() replaced dash with underscore.

AD7six committed:
Correct doc blocks

AD7six committed:
Fix bad whitespace

markstory commented:
Having underscore and dasherize work the same sounds good to me. What about delimit() instead of normalize?

AD7six committed:
Make underscore and dasherize consistent

Also Rename normalize to delimit, it isn't normalizing (take
inconsistent and make consistent).

AD7six committed:
It's the Inflector test case jim

There's no need to put "Inflector" in the test method names

ADmad commented:
:+1:

AD7six committed:
Update docs and variable names

Use delimiter rather than replacement where appropriate, and update doc
blocks to not refer to strings rather than words or word groups - which
isn't that clear.
